Add new question type 'multiple' (#19)
ministryofjustice/hmpps-book-secure-move-frameworks: frameworks/person-escort-record/README.md

### Question options

> - `type` **(required)** - the type of question input. Current supported types: `radio`, `checkbox`, `text`, `textarea`, `multiple`

Reviewer: I know we've been struggling a bit with the name of this. What about `add_multiple` - similar to the [Home Office pattern](https://home-office-digital-patterns.herokuapp.com/patterns/add-multiple-things)? Or something along the lines of _repeating_? `repeating_questions`? Not sure though...

I think `multiple` on its own could be a little ambiguous. Could mean multiple responses...multiple options...

Reviewer: `question_loop`?

> - `type` **(required)** - the [type of question](#question-types) input. Current supported types: `radio`, `checkbox`, `text`, `textarea`, `add_multiple`

Reviewer: I wonder if we drop the second sentence and describe the types below to avoid needing to duplicate them?

- `question` **(required)** - the full text of the question, usually including a question mark, that should be displayed in forms and summary tables.

> - `list_item_name` - text to display in the "Add another ..." button and item names, for questions with type [`add_multiple`](#question-types)

Reviewer: ```suggestion
- `list_item_name` - text to display in the "Add another ..." button and item names, for questions with type [`add_multiple`](#question-types). Should be defined as a singular noun.
```

- `questions` - list of questions to display relating to a single item, for questions with type [`add_multiple`](#question-types)
- `hint` - hint text to display after the question, usually for advice on how to format an answer. Supports [markdown](#markdown-support).
- `description` - text to display as the summary table row heading for this question. If not supplied, will use `question` value
- `display` - list of further display styles for question input
  - `unit` - a one-character string positioned at the beginning of a textbox (currency or number entry textboxes)

>   - `width` - the width of a question input

Author: add more info re width measurements

Author: @teneightfive need row and width? (character width vs percentage or both)

>     - `label` **(required)** - text displayed on the text box label

Reviewer: ```suggestion
    - `type` - the [type of question](#question-types) input. Current supported types for followup comments: `text`, `textarea`
    - `label` **(required)** - text displayed on the text box label
```

    - `hint` - hint text to display after the label, usually for advice on how to format an answer. Supports [markdown](#markdown-support).
    - `display` - list of further display styles for question input

>       - `rows` - the number of rows to display for `textarea` questions. If set to `1` will display a `text` element instead of a `textarea` element.

Reviewer: ```suggestion
      - `rows` - the number of rows to display for `textarea` questions
```

- `validations` - list of [validations](#question-validation) that need to be applied to this question
  - `type` **(required)** - the validation error key
  - `message` - custom text that will be displayed in the form
  - `value` - further data required for some [validation types](#question-validation).

> #### Multiple type questions

Reviewer: What about using `#### Questions types` as the heading, similar to the `Question validations` heading below? And it allows us to use this section to describe any types that might need more explanation.

- `validations` - list of [validations](#question-validation) that need to be applied to this question
  - `type` **(required)** - the validation error key
  - `message` - custom text that will be displayed in the form
  - `value` - further data required for some [validation types](#question-validation).

> #### Question types

Reviewer: ```suggestion
#### Question types

```

> A question can support multiple types of input: (`radio`), (`checkbox`), (`text`), (`textarea`) and (`add_multiple`).

Reviewer: ```suggestion
Supported types:

- `text` - renders a text form element and accepts a response as a string
- `textarea` - renders a text form element and accepts a response as a string
- `radio` - renders radio form elements and accepts a response as either a string or an object. Radio items are defined under `options`.
- `checkbox` - renders checkbox form elements and accepts a response as either an array or a collection. Checkbox items are defined under `options`.
- `add_multiple` - allows a set of questions to be grouped and repeated within a block and accepts a response as a collection. Child questions to display in each group are defined under `questions`. The noun for the item is defined under `list_item_name` in singular form.
```

Questions with type (`add_multiple`) are defined as a question that allows adding variations of responses for a group of the same (`questions`) multiple times, grouped under a (`list_item_name`). 

> The supported [validations](#question-validation) for this question allow setting the minimum and maximum number of items allowed, using the (`minlength`) and (`maxlength`) options.

Reviewer: ```suggestion
The supported [validations](#question-validation) for this question allow setting the minimum and maximum number of items allowed, using `minlength` and `maxlength` validations.
```

- `validations` - list of [validations](#question-validation) that need to be applied to this question
  - `type` **(required)** - the validation error key
  - `message` - custom text that will be displayed in the form

>   - `value` - further data required for some [validation types](#question-validation).

Reviewer: ```suggestion
  - `arguments` - array of validator arguments required for some [validation types](#question-validation), eg. `{ type: 'minlength', arguments: [24] }`
```

Totally stealing the idea from the [HO Form wizard](https://github.com/UKHomeOffice/passports-form-wizard#field-options)...

- `required` - mark a question as mandatory. For radios or checkboxes it will require at least one option to be selected, for text or textarea questions it will require at least 1 character

> - `min_length` - mark the minimum length of items required for a question with a response of type list. This validation requires an extra `value` option which specifies the length as an integer. If not specified defaults to `0`.

Reviewer: Could we remove the `_` from these? It's a little FE selfish but it means I don't need to do extra transformation when passing through to the form wizard and using [the validations](https://github.com/UKHomeOffice/passports-form-wizard/blob/9dd5356e7d62c3c4050ed83e43a8a2c543577429/lib/validation/validators.js#L29) from there that are already defined.

Reviewer: Maybe we should also mention in this validations section that it is only support on particular question types.

Author: Amended, re the particular question types, I've added `question with a response of type list` to allow this to be used for any questions which could have a list of responses instead of specifying `add_multiple`. In the question type question I do mention those validations
